<?xml version="1.0" ?>


<robot name="leg" xmlns:xacro="http://www.ros.org/wiki/xacro">


    <gazebo>
        <plugin filename="libgazebo_ros_control.so" name="gazebo_ros_control">
            <robotNamespace>/leg</robotNamespace>
            <robotSimType>
                gazebo_ros_control/DefaultRobotHWSim
            </robotSimType>
        </plugin>
    </gazebo>


    <material name="black">
        <color rgba="0.0 0.0 0.0 1.0"/>
    </material>
    <material name="blue">
        <color rgba="0.0 0.0 0.8 1.0"/>
    </material>
    <material name="green">
        <color rgba="0.0 0.8 0.0 1.0"/>
    </material>
    <material name="grey">
        <color rgba="0.2 0.2 0.2 1.0"/>
    </material>
    <material name="silver">
        <color rgba="0.913725490196 0.913725490196 0.847058823529 1.0"/>
    </material>
    <material name="orange">
        <color rgba="1.0 0.423529411765 0.0392156862745 1.0"/>
    </material>
    <material name="brown">
        <color rgba="0.870588235294 0.811764705882 0.764705882353 1.0"/>
    </material>
    <material name="white_gray">
        <color rgba="0.823529411765 0.823529411765 0.823529411765 1.0"/>
    </material>
    <material name="gold">
        <color rgba="0.811764705882 0.678431372549 0.301960784314 1.0"/>
    </material>
    <material name="white_blue">
        <color rgba="0.462745098039 0.8 0.992156862745 1.0"/>
    </material>
    <material name="white_black">
        <color rgba="0.494117647059 0.494117647059 0.494117647059 1.0"/>
    </material>
    <material name="red">
        <color rgba="0.8 0.0 0.0 1.0"/>
    </material>
    <material name="white">
        <color rgba="1.0 1.0 1.0 1.0"/>
    </material>

    <gazebo refrence="trunk">
        <mu1>0.9</mu1>
        <mu2>0.9</mu2>
        <material>Gazebo/DarkGrey</material>
    </gazebo>

    <gazebo refrence="jump">
        <mu1>0.9</mu1>
        <mu2>0.9</mu2>
        <material>Gazebo/DarkGrey</material>
    </gazebo>

    <gazebo reference="hip">
        <mu1>0.9</mu1>
        <mu2>0.9</mu2>
        <material>Gazebo/Orange</material>
  </gazebo>
  <gazebo reference="thigh">
        <mu1>0.9</mu1>
        <mu2>0.9</mu2>
        <self_collide>1</self_collide>
        <!-- <material>Gazebo/White</material> -->
        <kp value="1000000.0"/>
        <kd value="1.0"/>
  </gazebo>
  <gazebo reference="calf">
        <mu1>0.9</mu1>
        <mu2>0.9</mu2>
        <self_collide>1</self_collide>
      <material>Gazebo/DarkGray</material>
  </gazebo>
  <gazebo reference="foot">
        <mu1>0.9</mu1>
        <mu2>0.9</mu2>
        <self_collide>1</self_collide>
        <material>Gazebo/DarkGray</material>
        <kp value="1000000.0"/>
        <kd value="1.0"/>
  </gazebo>

    <link name="world"/>
    <gazebo reference="world">
        <static>true</static>
    </gazebo>

    <link name="trunk">
        <visual>
        <origin rpy="0 0 0" xyz="0 0 0.045"/>
        <geometry>
            <box size="0.075 0.06 0.09"/>
        </geometry>
        <material name="white"/>
        </visual>
        <collision>
            <origin rpy="0 0 0" xyz="0 0 0.045"/>
            <geometry>
                <box size="0.075 0.06 0.09"/>
            </geometry>
        </collision>
        <inertial>
            <origin rpy="0 0 0" xyz="0 0 0.045"/>
            <mass value="0.001"/>
            <inertia ixx="0.011253" ixy="0" ixz="0.0" iyy="0.036203" iyz="-0.0" izz="0.042673"/>
            </inertial>
        </link>

    <joint name="trunk_to_world" type="fixed">
        <origin rpy="0 0 0" xyz="0 0 0.0"/>
        <parent link="world"/>
        <child link="trunk"/>
    </joint>

    

    

    <!-- <joint name="jumper" type="prismatic">
        <parent link="trunk"/>
        <axis xyz="0 0 1"/>
        <child link="jump"/>
        <limit effort="20" lower="-0.2" upper="0.5" velocity="0.5"/>
        <origin rpy="0 0 0" xyz="0.15 0.07 0.9"/>
        <dynamics damping="0.1" friction="1"/>
    </joint> -->


<!--    <joint name="jumper" type="prismatic">-->
<!--        <parent link="trunk"/>-->
<!--        <axis xyz="0 0 1"/>-->
<!--        <child link="jump"/>-->
<!--        <limit effort="20" lower="-0.2" upper="0.5" velocity="0.5"/>-->
<!--        <origin rpy="0 0 0" xyz="0.15 0.07 0.9"/>-->
<!--        <dynamics damping="0.1" friction="1"/>-->
<!--    </joint>-->


<!--    <link name="jump">-->
<!--        <visual>-->
<!--            <origin rpy="3.14159265359 0 0" xyz="0.01 0 0"/>-->
<!--            <geometry>-->
<!--                <box size="0.02 0.08 0.093"/>-->
<!--            </geometry>-->
<!--            <material name="black"/>-->
<!--        </visual>-->
<!--        <collision>-->
<!--            <origin rpy="3.14159265359 0 0" xyz="0.01 0 0"/>-->
<!--            <geometry>-->
<!--                <box size="0.02 0.08 0.093"/>-->
<!--            </geometry>-->
<!--        </collision>-->
<!--        <inertial>-->
<!--            <origin rpy="0 0 0" xyz="0.01 0 0"/>-->
<!--            <mass value="0.002"/>-->
<!--            <inertia ixx="0.0007148275" ixy="0.0" ixz="0.0" iyy="0.0007955775" iyz="0.0" izz="0.00068875"/>-->
<!--        </inertial>-->
<!--    </link>-->

    <joint name="hip_joint" type="revolute">
        <origin rpy="0 0 0" xyz="0.025 0 0.0"/>
        <parent link="trunk"/>
        <child link="hip"/>
        <axis xyz="1 0 0"/>
        <dynamics damping="0.1" friction="0.1"/>
        <limit effort="20" lower="-3.14" upper="3.14" velocity="1.6"/>
  </joint>

  <link name="hip">
    <visual>
      <origin rpy="0 0 0" xyz="0.03 0.0 0.0"/>
      <geometry>
        <box size="0.06 0.08 0.093"/>
      </geometry>
      <material name="gold"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0.03 0.0 0.0"/>
      <geometry>
        <box size="0.06 0.08 0.093"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0.03 0.0 0.0"/>
      <mass value="0.63"/>
      <inertia ixx="0.0007148275" ixy="0.0" ixz="0.0" iyy="0.0007955775" iyz="0.0" izz="0.00068875"/>
    </inertial>
  </link>

    <joint name="thigh_joint" type="revolute">
    <origin rpy="0 0 0" xyz="0.06 0.015 0"/>
    <parent link="hip"/>
    <child link="thigh"/>
    <axis xyz="0 1 0"/>
    <dynamics damping="0.1" friction="0.1" />
    <limit effort="20" lower="-0.523598775598" upper="3.92699081699" velocity="1.6"/>
  </joint>

  <link name="thigh">
    <visual>
      <origin rpy="0 0 0" xyz="0 0.06 -0.105915"/>
      <geometry>
        <box size="0.06 0.07125 0.21183"/>
      </geometry>
      <material name="white_blue"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0.06 -0.105915"/>
      <geometry>
        <box size="0.06 0.07125 0.21183"/>
      </geometry>
      <material name="white_blue"/>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0.06 -0.105915"/>
      <mass value="1.062"/>
      <inertia ixx="0.00261820866667" ixy="0.0" ixz="0.0" iyy="0.00265482216667" iyz="0.0" izz="0.000158764166667"/>
    </inertial>
  </link>
  <!-- <gazebo reference="thigh">
    <material>ArucoVisualMarker0/Marker</material>
  </gazebo> -->
  
  <!-- //////////////////////calf//////////////////// -->
  <joint name="calf_joint" type="revolute">
    <origin rpy="0 0 0" xyz="0 0.07125 -0.21183"/>
<!--    <origin rpy="0 0 0" xyz="0 0.01 -0.1881"/>-->
    <parent link="thigh"/>
    <child link="calf"/>
    <axis xyz="0 1 0"/>
    <dynamics damping="0.1" friction="0.1"/>
    <limit effort="20" lower="-3.14159265359" upper="3.14159265359" velocity="1.6"/>
  </joint>
  <link name="calf">
    <visual>
      <origin rpy="0 0 0" xyz="0.0 0.01 -0.120"/>
      <geometry>
        <box size="0.03 0.02 0.240"/>
          <!--mesh filename="package://first_leg/mesh/calf.dae" scale="1 1 1"/-->
      </geometry>
      <material name="white_blue"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0.0 0.01 -0.120"/>
      <geometry>
          <box size="0.03 0.02 0.240"/>
        <!--mesh filename="package://first_leg/mesh/Sim_0_00_04.STL" scale="1 1 1"/-->
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0.0 0.01 -0.120"/>
      <mass value="0.133"/>
      <inertia ixx="0.000214698666667" ixy="0.0" ixz="0.0" iyy="0.000214698666667" iyz="0.0" izz="2.73066666667e-06"/>
    </inertial>
  </link>
  
  <!-- //////////////////////foot//////////////////// -->
  <joint name="foot_fixed" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 -0.240"/>
<!--    <origin rpy="0 0 0" xyz="0 0.01 -0.1881"/>-->
    <parent link="calf"/>
    <child link="foot"/>
    <axis xyz="0 1 0"/>
    <dynamics damping="0.1" friction="0.1"/>
    <limit effort="20" lower="-3.14159265359" upper="3.14159265359" velocity="10.6"/>
  </joint>

  <link name="foot">
    <visual>
      <origin rpy="0 0 0" xyz="0 0.002 0"/>
      <geometry>
        <sphere radius="0.025"/>
      </geometry>
      <material name="white_black"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0.002 0"/>
      <geometry>
        <sphere radius="0.025"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0.002 0"/>
      <mass value="0.005"/>
      <inertia ixx="4.2135e-05" ixy="0.0" ixz="0.0" iyy="4.2135e-05" iyz="0.0" izz="4.2135e-05"/>
    </inertial>
  </link>

  <joint name="camera_joint" type="fixed">
    <origin rpy="0 0 -1.57" xyz="0 0.6 -0.2"/>
    <parent link="world"/>
    <child link="camera_link"/>
  </joint>

  <link name="camera_link">
    <inertial>
      <mass value="0.200"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="5.8083e-4" ixy="0" ixz="0" iyy="3.0833e-5" iyz="0" izz="5.9083e-4"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://hector_sensors_description/meshes/asus_camera/asus_camera_simple.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://hector_sensors_description/meshes/asus_camera/asus_camera_simple.dae"/>
      </geometry>
    </collision>
  </link>
  <joint name="camera_depth_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0.0 0.049 0.0"/>
    <parent link="camera_link"/>
    <child link="camera_depth_frame"/>
  </joint>
  <link name="camera_depth_frame"/>
  <joint name="camera_depth_optical_joint" type="fixed">
    <origin rpy="-1.57079632679 0.0 -1.57079632679" xyz="0 0 0"/>
    <parent link="camera_depth_frame"/>
    <child link="camera_depth_optical_frame"/>
  </joint>
  <link name="camera_depth_optical_frame"/>
  <joint name="camera_rgb_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0.0 0.022 0.0"/>
    <parent link="camera_link"/>
    <child link="camera_rgb_frame"/>
  </joint>
  <link name="camera_rgb_frame"/>
  <joint name="camera_rgb_optical_joint" type="fixed">
    <origin rpy="-1.57079632679 0.0 -1.57079632679" xyz="0 0 0"/>
    <parent link="camera_rgb_frame"/>
    <child link="camera_rgb_optical_frame"/>
  </joint>
  <link name="camera_rgb_optical_frame"/>
  <gazebo reference="camera_depth_frame">
    <sensor name="camera" type="depth">
      <update_rate>20</update_rate>
      <camera>
        <horizontal_fov>1.09606677025</horizontal_fov>
        <image>
          <format>R8G8B8</format>
          <width>640</width>
          <height>480</height>
        </image>
        <clip>
          <near>0.5</near>
          <far>9</far>
        </clip>
      </camera>
      <plugin filename="libgazebo_ros_openni_kinect.so" name="camera_camera_controller">
        <imageTopicName>camera/rgb/image_raw</imageTopicName>
        <cameraInfoTopicName>camera/rgb/camera_info</cameraInfoTopicName>
        <depthImageTopicName>camera/depth/image_raw</depthImageTopicName>
        <depthImageCameraInfoTopicName>camera/depth/camera_info</depthImageCameraInfoTopicName>
        <pointCloudTopicName>camera/depth/points</pointCloudTopicName>
        <frameName>camera_depth_optical_frame</frameName>
      </plugin>
    </sensor>
  </gazebo>

  <!--/////////////// transmission tags ////////////////////// -->
  <transmission name="hip_tran">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="hip_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="hip_motor">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="thigh_tran">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="thigh_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="thigh_motor">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="calf_tran">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="calf_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="calf_motor">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
</robot>
